### PR TITLE
Add Kubernetes(Collector)

### PR DESCRIPTION
**Pull request by DazWilkin**
What do you think to my progressing this?

It provides basic (cluster, nodepool, node) counts for Digital Ocean's Kubernetes service.

```bash
digitalocean_errors_total{collector="kubernetes"} 0
# HELP digitalocean_kubernetes_cluster_count Number of Kubernetes clusters
# TYPE digitalocean_kubernetes_cluster_count gauge
digitalocean_kubernetes_cluster_count 1
# HELP digitalocean_kubernetes_cluster_up If 1 the kubernetes cluster is up and running, 0 otherwise
# TYPE digitalocean_kubernetes_cluster_up gauge
digitalocean_kubernetes_cluster_up{id="[[REDACTED]]",name="[[REDACTED]]",region="sfo2"} 0
# HELP digitalocean_kubernetes_nodepools_count Number of Kubernetes nodepools
# TYPE digitalocean_kubernetes_nodepools_count gauge
digitalocean_kubernetes_nodepools_count{id="[[REDACTED]]",name="[[REDACTED]]",region="sfo2"} 1
# HELP digitalocean_kubernetes_nodes_count Number of Kubernetes nodes
# TYPE digitalocean_kubernetes_nodes_count gauge
digitalocean_kubernetes_nodes_count{id="[[REDACTED]]",name="[[REDACTED]]",region=""} 3
```

By the way, this inspired a similar solution for Linode ([link](https://github.com/DazWilkin/linode-exporter)). My [motivation](https://pretired.dazwilkin.com/posts/191218/) is to standardize on a way to monitor and alert on resources I'm using across multiple cloud platforms (Google Cloud Platform, Digital Ocean and Linode) in order to keep my bills reasonable.

Since you created this, Go Modules are now a good package management solution for Golang.If you're supportive, I propose replacing dep with Go Modules?

### CODE REVIEW

**Comment by metalmatze** [comment]
Thank you for starting this.
I'm both a users of Digitalocean's Kubernetes offering and use Go modules everywhere now. So I'm 100% positive on continuing with this! 😊 💯

File: collector/kubernetes.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		ch <- prometheus.MustNewConstMetric(
+			c.NodePools,
+			prometheus.GaugeValue,
+			float64(len(cluster.NodePools)),
+			labels...,
+		)
+		var wg sync.WaitGroup
+		for _, nodepool := range cluster.NodePools {
+			wg.Add(1)
+			go func(np *godo.KubernetesNodePool) {
```
Comment (metalmatze) [inline collector/kubernetes.go]: We don't really need this goroutine here, do we? It should just work without it and then we don't need the working group either.
Am I missing something?

Comment (DazWilkin) [reply collector/kubernetes.go]: I was anticipating having more work to do calculating metrics for the NodePools; I'll remove.

Comment (DazWilkin) [reply collector/kubernetes.go]: Resolved: https://github.com/metalmatze/digitalocean_exporter/pull/5/commits/4b6ac4c31acf1db7d6b9ee1cf6a54119a4b91b6f

**Comment by DazWilkin** [comment]
I think this is a suitable basic implementation.

File: collector/kubernetes.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+			c.NodePools,
+			prometheus.GaugeValue,
+			float64(len(cluster.NodePools)),
+			labels...,
+		)
+
+		for _, nodepool := range cluster.NodePools {
+			// Assume NodePools are constrained to the cluster's Region
+			// If so, we can labels a cluster's NodePools by the cluster's region
+			go func(region string, np *godo.KubernetesNodePool) {
```
Comment (metalmatze) [inline collector/kubernetes.go]: I need to double check what the underlying implementation is and if this channel is buffered. But s
Nevertheless if we don't use the wait group we also can't use a goroutine here. 
I'm reviewing this on my phone, so I think it should work without the goroutine, but I need to verify that later.

File: collector/kubernetes.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	NodePools *prometheus.Desc
+	Nodes     *prometheus.Desc
+}
+
+// NewKubernetesCollector returns a new KubernetesCollector
+func NewKubernetesCollector(logger log.Logger, errors *prometheus.CounterVec, client *godo.Client, timeout time.Duration) *KubernetesCollector {
+	errors.WithLabelValues("kubernetes").Add(0)
+
+	// Version refers to the upstream Kubernetes version as well as the DigitalOcean revision
+	clusterLabels := []string{"id", "name", "region", "version"}
```
Comment (metalmatze) [inline collector/kubernetes.go]: I'm actually not sure about the version. It's most likely fine, but I feel like most people get this via kube-state-metrics anyway.
WDYT?

File: collector/kubernetes.go (hunk 2 above)
Comment (DazWilkin) [reply collector/kubernetes.go]: Crap, sorry. My mistake.

Resolved: https://github.com/metalmatze/digitalocean_exporter/pull/5/commits/b9b69c3de21556352cb0ea8900dcec95b7b39152

**Comment by metalmatze** [comment]
Alright. One last comment about the metric names.
The `_count` suffix is usually not used in Prometheus metrics, we can just leave it away as it's implicit.
```diff
- digitalocean_kubernetes_cluster_count 1
+ digitalocean_kubernetes_clusters 1
digitalocean_kubernetes_cluster_up{id="f74f45ec-0c53-4e96-8f22-XXX",name="metalmatze",region="fra1",version="1.14.8-do.2"} 1
- digitalocean_kubernetes_nodepools_count{id="f74f45ec-0c53-4e96-8f22-XXX",name="metalmatze",region="fra1",version="1.14.8-do.2"} 1
+ digitalocean_kubernetes_nodepools{id="f74f45ec-0c53-4e96-8f22-XXX",name="metalmatze",region="fra1",version="1.14.8-do.2"} 1
- digitalocean_kubernetes_nodes_count{id="5879b3c5-1b25-49d3-9ff3-XXX",name="pool-0dgh51ahd",region="fra1"} 3
+ digitalocean_kubernetes_nodes{id="5879b3c5-1b25-49d3-9ff3-XXX",name="pool-0dgh51ahd",region="fra1"} 3
```

If you want to, you can rebase against the new current master to make use of Go modules and the activated Drone CI, otherwise, we can simply merge it too, as I verified it locally. :blush:

**Comment by DazWilkin** [comment]
I realized my mistake in computing `_count` too.

I will remove the metric entirely.

Count should be computed by summing the e.g. `_up` metrics.

**Comment by metalmatze** [comment]
Ready when you are. Is this good to go from your side?

**Comment by metalmatze** [comment]
Thanks a lot for this valuable contribution! :tada:

**Pull request merged**